File: backend/optional_upgrades.py
"""Suggest optional ingredient upgrades not in the user's pantry."""

from __future__ import annotations

from ingredient_relevance import ingredients_match, normalize_ingredient

UPGRADE_CATALOG: list[tuple[str, str, str, str]] = [
    ("butter", "חמאה", "butter", "Adds richness and helps bind the mixture"),
    ("vanilla", "וניל", "vanilla", "Enhances sweetness and aroma"),
    ("oil", "שמן", "oil", "Helps prevent sticking and improves texture"),
    ("flour", "קמח", "flour", "Adds structure for baking"),
    ("baking powder", "אבקת אפייה", "baking powder", "Helps desserts rise and stay light"),
    ("egg", "ביצה", "egg", "Binds ingredients and improves texture"),
    ("milk", "חלב", "milk", "Softens and enriches the mixture"),
    ("cocoa", "אבקת קקאו", "cocoa powder", "Deepens chocolate flavor"),
    ("honey", "דבש", "honey", "Adds natural sweetness and moisture"),
    ("salt", "מלח", "salt", "Balances sweetness and enhances flavor"),
]

UPGRADE_REASONS_HE: dict[str, str] = {
    "butter": "מוסיפה עשירות ומחברת את התערובת",
    "vanilla": "מעצימה את הניחוח והמתיקות",
    "oil": "מונע הידבקות ומשפר מרקם",
    "flour": "מוסיפה מבנה לאפייה",
    "baking powder": "מסייעת לקינוחים להתרומם",
    "egg": "מחברת מרכיבים ומשפרת מרקם",
    "milk": "מרככת ומעשירה את התערובת",
    "cocoa": "מעמיקה טעם שוקולד",
    "honey": "מוסיפה מתיקות טבעית ולחות",
    "salt": "מאזנת מתיקות ומבליטה טעמים",
}


def build_optional_upgrades(
    user_ingredients: list[str],
    *,
    language: str = "he",
    recipe_type: str = "meal",
    limit: int = 3,
) -> list[dict[str, str]]:
    """Return up to `limit` optional upgrades not already in the user's list."""
    normalized_user = {normalize_ingredient(item) for item in user_ingredients if item}
    upgrades: list[dict[str, str]] = []

    for canon, he_label, en_label, _en_reason in UPGRADE_CATALOG:
        if any(ingredients_match(canon, user) for user in user_ingredients):
            continue
        if canon in normalized_user:
            continue
        if recipe_type == "dessert" and canon in {"salt"} and language == "he":
            pass  # salt still useful in desserts

        label = he_label if language == "he" else en_label
        reason = UPGRADE_REASONS_HE.get(canon, _en_reason)
        if language == "en":
            reason = _en_reason

        upgrades.append({"ingredient": label, "reason": reason})
        if len(upgrades) >= limit:
            break

    return upgrades

File: src/utils/optionalUpgrades.js
import { canonicalIngredient, ingredientsMatch } from '../data/ingredientKnowledge'

/**
 * Savory finishing upgrades. `match` lists user-ingredient canons this pairs
 * especially well with, so suggestions stay relevant to the actual dish.
 */
const SAVORY_UPGRADES = [
  { canon: 'parsley', he: 'פטרוזיליה קצוצה', en: 'chopped parsley', reasonHe: 'מוסיפה רעננות וצבע למנה', reasonEn: 'Adds freshness and color', match: [] },
  { canon: 'chili flakes', he: 'פתיתי צ׳ילי', en: 'chili flakes', reasonHe: 'מוסיפים חריפות עדינה', reasonEn: 'Add a gentle kick of heat', match: [] },
  { canon: 'lemon', he: 'מיץ לימון', en: 'a squeeze of lemon', reasonHe: 'מאזן ומבהיר את הטעמים', reasonEn: 'Brightens and balances the flavors', match: [] },
  { canon: 'parmesan', he: 'פרמזן מגורר', en: 'grated parmesan', reasonHe: 'מוסיף מליחות ועומק לרוטב ולפסטה', reasonEn: 'Adds salty depth to pasta and sauce', match: ['pasta', 'cream', 'cheese'] },
  { canon: 'nutmeg', he: 'אגוז מוסקט', en: 'nutmeg', reasonHe: 'מעדן ומעשיר רטבי שמנת', reasonEn: 'Rounds out and enriches cream sauces', match: ['cream'] },
  { canon: 'oregano', he: 'אורגנו', en: 'oregano', reasonHe: 'מדגיש את טעם העגבניות והגבינה', reasonEn: 'Highlights the tomato and cheese', match: ['tomato', 'cheese'] },
  { canon: 'basil', he: 'בזיליקום טרי', en: 'fresh basil', reasonHe: 'משתלב מצוין עם עגבניות וגבינה', reasonEn: 'Pairs beautifully with tomato and cheese', match: ['tomato', 'cheese'] },
  { canon: 'olives', he: 'זיתים', en: 'olives', reasonHe: 'מוסיפים מליחות ים-תיכונית', reasonEn: 'Add briny Mediterranean flavor', match: ['tomato', 'cheese'] },
  { canon: 'green onion', he: 'בצל ירוק', en: 'green onion', reasonHe: 'מוסיף רעננות וטעם עדין', reasonEn: 'Adds a fresh, mild onion note', match: ['egg', 'cheese'] },
  { canon: 'feta', he: 'גבינת פטה', en: 'feta cheese', reasonHe: 'מוסיפה מליחות וטעם עשיר', reasonEn: 'Adds a salty, tangy bite', match: ['tomato', 'cheese', 'egg'] },
  { canon: 'mushrooms', he: 'פטריות', en: 'mushrooms', reasonHe: 'מוסיפות מרקם ועומק אומאמי', reasonEn: 'Add texture and umami depth', match: ['egg'] },
]

/** Sweet finishing upgrades for desserts. */
const DESSERT_UPGRADES = [
  { canon: 'vanilla', he: 'תמצית וניל', en: 'vanilla extract', reasonHe: 'מעצימה ניחוח ומתיקות', reasonEn: 'Enhances aroma and sweetness', match: [] },
  { canon: 'cinnamon', he: 'קינמון', en: 'cinnamon', reasonHe: 'מוסיף חמימות וניחוח', reasonEn: 'Adds warmth and aroma', match: [] },
  { canon: 'chocolate', he: 'שבבי שוקולד', en: 'chocolate chips', reasonHe: 'מוסיפים עושר ומתיקות', reasonEn: 'Add richness and sweetness', match: [] },
  { canon: 'honey', he: 'דבש', en: 'honey', reasonHe: 'ממתיק בעדינות טבעית', reasonEn: 'Sweetens with a natural touch', match: [] },
  { canon: 'coconut', he: 'קוקוס', en: 'shredded coconut', reasonHe: 'מוסיף מרקם וטעם', reasonEn: 'Adds texture and flavor', match: [] },
]

const DAIRY_UPGRADE_CANONS = new Set(['feta', 'parmesan', 'cheese', 'mozzarella', 'ricotta'])

function userAllowsDairy(userIngredients, selectedCategory = 'any') {
  if (selectedCategory === 'dairy') return true
  return (userIngredients ?? []).some((user) => {
    const canon = canonicalIngredient(user)
    return canon && ['cheese', 'milk', 'cream', 'butter', 'yogurt', 'feta', 'parmesan'].includes(canon)
  })
}

function userHas(userIngredients, canon) {
  return (userIngredients ?? []).some(
    (user) => ingredientsMatch(user, canon) || canonicalIngredient(user) === canon,
  )
}

/**
 * Context-aware optional upgrades for fallback/mock recipes. Suggestions are
 * chosen from a savory or dessert catalog (by recipe type) and ranked by how
 * well they match the actual ingredients, so a savory egg/tomato/cheese dish
 * gets herbs and toppings — never vanilla or random oil.
 *
 * @param {string[]} userIngredients
 * @param {{ language?: string, recipeType?: string, limit?: number }} [options]
 */
export function buildOptionalUpgrades(
  userIngredients,
  { language = 'he', recipeType = 'meal', selectedCategory = 'any', limit = 3 } = {},
) {
  const catalog = recipeType === 'dessert' ? DESSERT_UPGRADES : SAVORY_UPGRADES
  const userCanons = new Set((userIngredients ?? []).map((user) => canonicalIngredient(user)).filter(Boolean))
  const allowDairy = userAllowsDairy(userIngredients, selectedCategory)

  return catalog
    .filter((item) => !userHas(userIngredients, item.canon))
    .filter((item) => allowDairy || !DAIRY_UPGRADE_CANONS.has(item.canon))
    .map((item) => ({
      item,
      score: (item.match ?? []).filter((canon) => userCanons.has(canon)).length,
    }))
    .sort((a, b) => b.score - a.score)
    .slice(0, limit)
    .map(({ item }) => ({
      ingredient: language === 'he' ? item.he : item.en,
      reason: language === 'he' ? item.reasonHe : item.reasonEn,
    }))
}
